typegate/src/typecheck/result.ts

```diff
@@ -0,0 +1,387 @@
+    if (errors.length > 0) {
+      const messages = errors.map(([path, msg]) => `  - at ${path}: ${msg}\n`)
+        .join("");
+      throw new Error(`Validation errors:\n${messages}`);
+    }
+  };
+}
+
+interface QueueEntry {
+  name: string;
+  typeIdx: number;
+  selectionSet?: SelectionSetNode | undefined;
+  path: string;
+}
+
+function validatorName(idx: number, counter: number | null) {
+  if (counter == null) return `validate_${idx}`;
+  else return `validate_${idx}_${counter}`;
+}
+
+export class ResultValidationCompiler {
+  codes: Map<string, string> = new Map();
+  counter = 0;
+
+  constructor(private tg: TypeGraph, private fragments: FragmentDefs) {}
+
+  private getRootQueueEntry(opDef: OperationDefinitionNode): QueueEntry {
+    const { name, operation, selectionSet } = opDef;
+    const rootTypeIdx = this.tg.type(0, Type.OBJECT).properties[operation];
+    if (rootTypeIdx == null) {
+      throw new Error(`Unsupported operation '${operation}'`);
+    }
+    const rootPath = name?.value ?? operation[0].toUpperCase();
+    // TODO check if selection set is required or prohibited
+    return {
+      name: validatorName(rootTypeIdx, ++this.counter),
+      typeIdx: rootTypeIdx,
+      selectionSet,
+      path: rootPath,
+    };
+  }
+
+  generate(opDef: OperationDefinitionNode) {
+    const cg = new CodeGenerator();
+    const rootEntry = this.getRootQueueEntry(opDef);
+    const queue: QueueEntry[] = [rootEntry];
+    const refs = new Set([rootEntry.name]);
+
+    for (
+      let entry = queue.shift();
+      entry != null;
+      entry = queue.shift()
+    ) {
+      refs.add(entry.name);
+      if (this.codes.has(entry.name)) {
+        continue;
+      }
+
+      const typeNode = this.tg.type(entry.typeIdx);
+
+      if (entry.name === "validate_typename") {
+        cg.generateStringValidator({
+          type: "string",
+          title: "__TypeName",
+          runtime: -1,
+          policies: [],
+        });
+      } else if (isScalar(typeNode)) {
+        if (entry.selectionSet != null) {
+          throw new Error(
+            `Unexpected selection set for scalar type '${typeNode.type}' at '${entry.path}'`,
+          );
+        }
+
+        if (typeNode.enum != null) {
+          cg.generateEnumValidator(typeNode);
+        } else {
+          switch (typeNode.type) {
+            case "boolean":
+              cg.generateBooleanValidator(typeNode);
+              break;
+            case "number":
+            case "integer":
+              cg.generateNumberValidator(typeNode);
+              break;
+            case "string":
+              cg.generateStringValidator(typeNode);
+              break;
+          }
+        }
+      } else {
+        // TODO: cannot check enum - perhaps we should disable enums for non-scalar types??
+
+        switch (typeNode.type) {
+          case "optional": {
+            const c = entry.selectionSet != null ? ++this.counter : null;
```

Comment:
I think that's a good idea.

Reply:
can we avoid ++ and use += 1 for clarity?

Reply:
> I don't quite follow, aren't `t.enum` variants `str` only ?

`t.enum()` is an alias to `t.string().enum()`. Any type is valid.